### PR TITLE
Make podIPs order match node IP family preference

### LINKED ISSUE #95251: DualStack: Ip family of status.podIP can be wrong depending on CNI-plugin
### Ip family of status.podIP can be wrong depending on CNI-plugin

**What happened**:

The `status.podIP` is taken from `status.podIPs[0]` without regard for the main family of the cluster. No warning or error is given if `status.podIP` gets the wrong family.

**What you expected to happen**:

That `status.podIP` is taken from the "main" ip family, or at the very least the POD is not started and an error is given if it's not.

**How to reproduce it (as minimally and precisely as possible)**:

Use the `bridge` CNI-plugin and `host-local` ipam. Use a single node cluster to avoid that `host-local` assign the same address to PODs in different nodes. This _may_ be CRI-plugin dependent, I use `cri-o/1.18.3`.

```
$ cat /etc/cni/net.d/10-bridge.conf 
{
  "cniVersion": "0.4.0",
  "name": "cni-x",
  "type": "bridge",
  "bridge": "cbr0",
  "isDefaultGateway": true,
  "hairpinMode": true,
  "ipam": {
    "type": "host-local",
    "ranges": [
      [ { "subnet": "1100::100/120" } ],
      [ { "subnet": "11.0.1.0/24" } ]
    ]
  }
}
$ ls /opt/cni/bin/
bridge*      host-local*    loopback*
```

The PODs are assigned dual addresses which has been supported since v1.9. Note the order; IPv6 first.

In a dual-stack _or a single-stack_ cluster with main family IPv4 `status.podIP` will get an IPv6 address even if the main family is IPv4;

```
$ kubectl get pod alpine-daemonset-4t6w5 -o json | jq .status.podIP
"1100::102"
```

**Anything else we need to know?**:

First, note that this is **not** a dual-stack problem. `status.podIP` may get the wrong family in a single-stack cluster if the PODs has dual addresses. The bug may however have been introduced with the dual-stack support, before `status.podIPs` K8s may have selected the correct family, I have not checked that.

This is reported and discussed in https://github.com/kubernetes/kubernetes/issues/94505. There it is considered as a CNI-plugin problem, the CNI-plugin must present addresses in the order K8s wants them. Problem is that CNI-plugins doesn't know this.

The addresses takes the path;
```
CNI-plugin -> CRI-plugin -> kubelet -> API-server
```
The problem can be addressed in either of these places.

The current situation is that the full responsibility lies on the `CNI-plugin`, it must send the addresses in the "correct order". This is not the best place. The CNI-plugins are not a K8s-only thing and to require that a "prefered-family" or something similar must be supported by CNI-plugins to be "Kubernetes compliant" should be avoided.

The CRI-plugin *is* a K8s-only thing and a "prefered-family" can be introduced so the CRI-plugin can sort the address array before sending it to `kubelet`. This is however undesirable since it adds a configuration complexity, it put a responsibility on the user or installation tool to configure the "prefered-family" to all CNI-plugins now and in the future.

The best is to handle this in K8s itself where the "main" family is known.
```
    // IP addresses allocated to the pod. This list
    // is inclusive, i.e. it includes the default IP address stored in the
    // "PodIP" field, and this default IP address must be recorded in the
    // 0th entry (PodIPs[0]) of the slice. The list is empty if no IPs have
    // been allocated yet.
    PodIPs []PodIP `json:"podIPs,omitempty" protobuf:"bytes,6,opt,name=podIPs"`
```

Is this really necessary? All communication works even if `podIPs[0]` is not of the main family. EndpointSlices seem to get this right regardless of order. Endpoint on the other hand seem to take the (old) podIP. So perhaps;
```
The "PodIP" field is set to the first address in podIPs that matches the main ip-family.
```
is sufficient?

If `podIP` *really must* be `podIPs[0]`, then K8s should sort the array so `podIPs[0]` belongs to the main family.

The problem exist on all supported K8s versions and on "master".

#### Current situation for dual-stack supporting CNI-plugins

If installed in the default way the situation is;

* Calico *always* send IPv4 first
* Cilum *always* send IPv6 first

I don't think you can control the order, but I have not asked.

**Environment**:
- Kubernetes versions: v1.17.12, v1.18.9, v1.19.2, v1.20.0-alpha.1, master v1.20.0-alpha.1.257+112dbd55860e60
- Cloud provider or hardware configuration: None
- OS (e.g: `cat /etc/os-release`): xcluster
- Kernel: linux-5.8.1
- Install tools: None
- Network plugin and version: bridge, host-local v0.8.7
- Others: CRI-plugin: cri-o 1.18.3

### PR DESCRIPTION
**Pull request by danwinship**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Kubelet currently sets `pod.status.podIPs` directly from the array of IPs returned from the runtime. But there are two problems with this:

1. `pod.status.podIPs` is restricted to be at most one IPv4 and at most one IPv6 IP, but the list of IPs from CRI/CNI has no such restriction; if the runtime returned, eg, 3 IPv4 addresses, then kubelet would try to set podIPs to that and then fail.
2. kubelet is assuming that the IPs are returned in preference order, but CRI/CNI does not actually specify that, and the consensus of the CNI maintainers in #95251 seems to be that we should *not* be assuming that; kubelet should treat the list as unordered, and impose its own logic on them. (As a concrete example of how this causes problems; Cilium always returns the IPv6 pod IP first if it is configured to do dual-stack, which currently means any Cilium dual-stack kubernetes cluster ends up having IPv6-primary pod IPs, whether the administrator wants that or not.) Although this technically doesn't break anything, it's considered ugly and confusing. (Maybe a bigger problem is that if the CNI plugin returns dual-stack pod IPs in a *single-stack* cluster, kubelet might end up grabbing a pod IP that is incompatible with the cluster configuration.)

This PR fixes kubelet to always pick an IP of the same family as its `--node-ip` for the first pod IP, and an IP of the opposite family (if any) for the second pod IP, and ignore any other IPs from the runtime.

**Which issue(s) this PR fixes**:
Fixes #95251

**Does this PR introduce a user-facing change?**:
```release-note
When a CNI plugin returns dual-stack pod IPs, kubelet will now try to respect the
"primary IP family" of the cluster by picking a primary pod IP of the same family
as the (primary) node IP, rather than assuming that the CNI plugin returned the IPs
in the order the administrator wanted (since some CNI plugins don't allow
configuring this).
```

/sig network
/priority important-longterm
/cc @uablrek @aojea @khenidak 
/auto-cc

### CODE REVIEW

**Comment by krmayankk** [comment]
So this issue should only be observed in dual stack configurations ?

**Comment by uablrek** [comment]
No, as @danwinship points out;

> (Maybe a bigger problem is that if the CNI plugin returns dual-stack pod IPs in a single-stack cluster, kubelet might end up grabbing a pod IP that is incompatible with the cluster configuration.)

So in a **single stack** cluster with primary family IPv4 where the CNI-plugin returns two addresses but the IPv6 address first you get;
```
  09:02:35 Server Version: v1.20.1
  09:02:35 CNI-plugin; bridge
  09:02:36 Single-stack cluster. BaseFamily=IPv4, phase.3 dual-stack API
$ kubectl get pod kahttp-6wqm8 -o wide
NAME           READY   STATUS    RESTARTS   AGE     IP          NODE     NOMINATED NODE   READINESS GATES
kahttp-6wqm8   1/1     Running   0          3m15s   1100::402   vm-004   <none>           <none>
```
**NOTE** this is a *single stack* IPv4 cluster with; `--feature-gates=IPv6DualStack=false,SCTPSupport=true`.

I use `bridge` CNI-plugin since it allows control of the order of addresses.

This is not as bad for K8s as it looks since `endpointslices` saves the day;
```
$ kubectl get endpoints kahttp
NAME     ENDPOINTS                                                     AGE
kahttp   [1100::104]:443,[1100::202]:443,[1100::303]:443 + 5 more...   4m42s
$ kubectl get endpointslices | grep kahttp
kahttp-lt8hr               IPv4          443,80           11.0.2.2,11.0.3.3,11.0.4.2 + 1 more...            5m
```
It is probably a disaster for other programs that looks at podIP and endpoints, for instance a service mesh.

**Comment by uablrek** [comment]
Same thing with this PR applied;
```
  09:23:36 Server Version: v1.21.0-alpha.0.786+a098b5b16c9a7f-dirty
  09:23:36 CNI-plugin; bridge
  09:23:36 Single-stack cluster. BaseFamily=IPv4, phase.3 dual-stack API
$ kubectl get pod kahttp-88cdx -o wide
NAME           READY   STATUS    RESTARTS   AGE    IP         NODE     NOMINATED NODE   READINESS GATES
kahttp-88cdx   1/1     Running   0          2m3s   11.0.2.3   vm-002   <none>           <none>
$ kubectl get endpoints kahttp
NAME     ENDPOINTS                                            AGE
kahttp   11.0.1.3:443,11.0.2.3:443,11.0.3.3:443 + 5 more...   2m18s
$ kubectl get endpointslices | grep kahttp
kahttp-48l5j               IPv4          443,80           11.0.2.3,11.0.4.2,11.0.3.3 + 1 more...            2m26s
```

**Review by uablrek — APPROVED** [review]
LGTM

**Comment by aojea** [comment]
> pull-kubernetes-verify — Job failed.

```
./hack/update-bazel.sh
./hack/update-gofmt.sh
```

it LGTM, should we document this somewhere else?
> kubelet is assuming that the IPs are returned in preference order, but CRI/CNI does not actually specify that, and the consensus of the CNI maintainers in #95251 seems to be that we should not be assuming that; kubelet should treat the list as unordered, and impose its own logic on them.

**Comment by danwinship** [comment]
> pull-kubernetes-verify — Job failed.

sigh, I always forget that

> it LGTM, should we document this somewhere else?

it's really a CNI thing, not a kubernetes thing...

**Comment by aojea** [comment]
> it's really a CNI thing, not a kubernetes thing...

I mean the kubelet behaviour 

> In dual-stack clusters, kubelet will now try to respect the "primary IP family" of the
> cluster when setting pod.status.podIPs, by picking a primary pod IP of the same family
> as the node IP rather than assuming that the CNI plugin returned the IPs in the order
> the administrator wanted.

**Comment by ehashman** [comment]
/retest
flake

I'd be +1 to documenting that the order output here matters and why.

File: pkg/kubelet/kubelet_pods_test.go [hunk 1]
```diff
@@ -2613,3 +2614,137 @@ func TestGenerateAPIPodStatusHostNetworkPodIPs(t *testing.T) {
 		})
 	}
 }
+
+func TestGenerateAPIPodStatusPodIPs(t *testing.T) {
```
Comment (ehashman) [inline pkg/kubelet/kubelet_pods_test.go:2618]: Great tests \o/

**Review by ehashman — COMMENTED** [review]
/lgtm

**Comment by krmayankk** [comment]
> No, as @danwinship points out;
> 
> > (Maybe a bigger problem is that if the CNI plugin returns dual-stack pod IPs in a single-stack cluster, kubelet might end up grabbing a pod IP that is incompatible with the cluster configuration.)
> 
> So in a **single stack** cluster with primary family IPv4 where the CNI-plugin returns two addresses but the IPv6 address first you get;
> 
> ```
>   09:02:35 Server Version: v1.20.1
>   09:02:35 CNI-plugin; bridge
>   09:02:36 Single-stack cluster. BaseFamily=IPv4, phase.3 dual-stack API
> $ kubectl get pod kahttp-6wqm8 -o wide
> NAME           READY   STATUS    RESTARTS   AGE     IP          NODE     NOMINATED NODE   READINESS GATES
> kahttp-6wqm8   1/1     Running   0          3m15s   1100::402   vm-004   <none>           <none>
> ```
> 
> **NOTE** this is a _single stack_ IPv4 cluster with; `--feature-gates=IPv6DualStack=false,SCTPSupport=true`.
> 
> I use `bridge` CNI-plugin since it allows control of the order of addresses.
> 
> This is not as bad for K8s as it looks since `endpointslices` saves the day;
> 
> ```
> $ kubectl get endpoints kahttp
> NAME     ENDPOINTS                                                     AGE
> kahttp   [1100::104]:443,[1100::202]:443,[1100::303]:443 + 5 more...   4m42s
> $ kubectl get endpointslices | grep kahttp
> kahttp-lt8hr               IPv4          443,80           11.0.2.2,11.0.3.3,11.0.4.2 + 1 more...            5m
> ```
> 
> It is probably a disaster for other programs that looks at podIP and endpoints, for instance a service mesh.
Cool thanks for the great reply. So you somehow used the bridge CNI plugin to force a different order in a single stack k8s deployment ?

File: pkg/kubelet/kubelet_pods_test.go [hunk 2]
```diff
@@ -2613,3 +2614,137 @@ func TestGenerateAPIPodStatusHostNetworkPodIPs(t *testing.T) {
...
+		t.Run(tc.name, func(t *testing.T) {
+			testKubelet := newTestKubelet(t, false /* controllerAttachDetachEnabled */)
+			defer testKubelet.Cleanup()
+			kl := testKubelet.kubelet
+			if tc.nodeIP != "" {
+				kl.nodeIPs = []net.IP{net.ParseIP(tc.nodeIP)}
+			}
+
+			pod := podWithUIDNameNs("12345", "test-pod", "test-namespace")
+			pod.Spec.HostNetwork = true
```
Comment (krmayankk) [inline pkg/kubelet/kubelet_pods_test.go:2732]: does this need to be true to test this ?

File: pkg/kubelet/kubelet_pods_test.go (hunk 1 above)
Comment (krmayankk) [reply pkg/kubelet/kubelet_pods_test.go:2618]: +1 great coverage

File: pkg/kubelet/kubelet_pods_test.go (hunk 2 above)
Comment (danwinship) [reply pkg/kubelet/kubelet_pods_test.go:2732]: ah, no, oops; I based this test off `TestGenerateAPIPodStatusHostNetworkPodIPs` but forgot to remove that line.

(It ends up having no effect on the test but it's confusing to have it there.)

**Comment by ehashman** [comment]
@krmayankk will leave this for you to throw an LGTM on

**Comment by uablrek** [comment]
@krmayankk 

> So you somehow used the bridge CNI plugin to force a different order in a single stack k8s deployment ?

Yes, snipplet from the start-script below. $i is a "node-index";

```bash
mkdir -p /etc/cni/net.d
if test "$IP_FAMILY_ORDER" = "64"; then
        cat > /etc/cni/net.d/10-bridge.conf <<EOF
{
  "cniVersion": "0.4.0",
  "name": "cni-x",
  "type": "bridge",
  "bridge": "cbr0",
  "isDefaultGateway": true,
  "hairpinMode": true,
  "ipam": {
    "type": "host-local",
    "ranges": [
      [ { "subnet": "1100::${i}00/120" } ],
      [ { "subnet": "11.0.$i.0/24" } ]
    ]
  }
}
EOF
else
        cat > /etc/cni/net.d/10-bridge.conf <<EOF
{
  "cniVersion": "0.4.0",
  "name": "cni-x",
  "type": "bridge",
  "bridge": "cbr0",
  "isDefaultGateway": true,
  "hairpinMode": true,
  "ipam": {
    "type": "host-local",
    "ranges": [
      [ { "subnet": "11.0.$i.0/24" } ],
      [ { "subnet": "1100::${i}00/120" } ]
    ]
  }
}
EOF
fi
```

**Comment by krmayankk** [comment]
> @krmayankk
> 
> > So you somehow used the bridge CNI plugin to force a different order in a single stack k8s deployment ?
> 
> Yes, snipplet from the start-script below. $i is a "node-index";
> 
> ```shell
> mkdir -p /etc/cni/net.d
> if test "$IP_FAMILY_ORDER" = "64"; then
>         cat > /etc/cni/net.d/10-bridge.conf <<EOF
> {
>   "cniVersion": "0.4.0",
>   "name": "cni-x",
>   "type": "bridge",
>   "bridge": "cbr0",
>   "isDefaultGateway": true,
>   "hairpinMode": true,
>   "ipam": {
>     "type": "host-local",
>     "ranges": [
>       [ { "subnet": "1100::${i}00/120" } ],
>       [ { "subnet": "11.0.$i.0/24" } ]
>     ]
>   }
> }
> EOF
> else
>         cat > /etc/cni/net.d/10-bridge.conf <<EOF
> {
>   "cniVersion": "0.4.0",
>   "name": "cni-x",
>   "type": "bridge",
>   "bridge": "cbr0",
>   "isDefaultGateway": true,
>   "hairpinMode": true,
>   "ipam": {
>     "type": "host-local",
>     "ranges": [
>       [ { "subnet": "11.0.$i.0/24" } ],
>       [ { "subnet": "1100::${i}00/120" } ]
>     ]
>   }
> }
> EOF
> fi
> ```

thanks you sir, this is very nice

**Comment by aojea** [comment]
pinging @mrunalp @mikebrow for Container Runtime feedback

**Review by aanm — APPROVED** [review]

**Comment by uablrek** [comment]
IMO this _must_ go into K8s v1.21 for dual-stack to go "beta". Enhancement freeze at Feb 9.

**Comment by aojea** [comment]
> IMO this _must_ go into K8s v1.21 for dual-stack to go "beta". Enhancement freeze at Feb 9.

totally agree

**Comment by aojea** [comment]
/assign @thockin 
for final approval

**Comment by derekwaynecarr** [comment]
the kubelet changes look good to me for sig-node.  ordering makes sense here.

/approve

**Review by thockin — COMMENTED** [review]
Need to rebase and nuke BUILD

**Comment by aojea** [comment]
/lgtm
it was already lgtmed before the rebase and there is a wide consensus

**Comment by uablrek** [comment]
Great :+1:

**Comment by Jean-Daniel** [comment]
For the record, this fix is only a partial fix, as it misses pod creation case: #103263

**Pull request merged**